pkg/common/workbuilder/workbuilder.go

```diff
@@ -0,0 +1,318 @@
+	return key, nil
+}
+
+func buildManifest(object runtime.Object) (workapiv1.Manifest, error) {
+	rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
+	if err != nil {
+		return workapiv1.Manifest{}, fmt.Errorf("failed to encode object %v, err: %v", object, err)
+	}
+	manifest := workapiv1.Manifest{RawExtension: runtime.RawExtension{Raw: rawObject}}
+	return manifest, nil
+}
+
+func generateRequiredManifestMapper(objects []runtime.Object) (requiredManifestMapper, error) {
+	var mapper = requiredManifestMapper{}
+	for _, object := range objects {
+		rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode object %v, err: %v", object, err)
+		}
+		key := manifestKey{
+			gvk: object.GetObjectKind().GroupVersionKind(),
+		}
+		accessor, err := meta.Accessor(object)
+		if err != nil {
+			return nil, err
+		}
+		key.namespace = accessor.GetNamespace()
+		key.name = accessor.GetName()
+		mapper[key] = workapiv1.Manifest{RawExtension: runtime.RawExtension{Raw: rawObject}}
+	}
+	return mapper, nil
+}
+
+func (w *WorkBuilder) compareManifests(existingWorks []workapiv1.ManifestWork,
+	requiredMapper requiredManifestMapper) (appliedWorks, deletedWorks []workapiv1.ManifestWork, err error) {
+	var updatedWorks []workapiv1.ManifestWork
+	for workIndex := 0; workIndex < len(existingWorks); workIndex++ {
+		work := existingWorks[workIndex].DeepCopy()
+		work.Spec.Workload.Manifests = []workapiv1.Manifest{}
+		for manifestIndex := 0; manifestIndex < len(existingWorks[workIndex].Spec.Workload.Manifests); manifestIndex++ {
+			manifest := existingWorks[workIndex].Spec.Workload.Manifests[manifestIndex]
+			key, err := generateManifestKey(manifest)
+			if err != nil {
+				return nil, nil, err
+			}
+			if _, ok := requiredMapper[key]; ok {
+				work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, requiredMapper[key])
+				delete(requiredMapper, key)
+				continue
+			}
+		}
+		updatedWorks = append(updatedWorks, *work)
+	}
+
+	// there are new added manifests
+	if len(requiredMapper) != 0 {
+		var createdManifests []workapiv1.Manifest
+		for _, manifest := range requiredMapper {
+			createdManifests = append(createdManifests, manifest)
+		}
+
+		// sort from big to small by size
+		sort.SliceStable(createdManifests, func(i, j int) bool {
+			return createdManifests[j].Size() < createdManifests[i].Size()
+		})
+
+		// sort from small to big by total size
+		sort.SliceStable(updatedWorks, func(i, j int) bool {
+			iTotalSize := totalSizeOfManifests(updatedWorks[i].Spec.Workload.Manifests)
+			jTotalSize := totalSizeOfManifests(updatedWorks[j].Spec.Workload.Manifests)
+			return jTotalSize > iTotalSize
+		})
+
+		var newWorkManifests []workapiv1.Manifest
+		for manifestIndex := 0; manifestIndex < len(createdManifests); manifestIndex++ {
+			appended := false
+			for workIndex := 0; workIndex < len(updatedWorks); workIndex++ {
```

Comment:
it does not look correct. I think you can sort the createdManifest at first

In the for loop, you would also need to find the min size existing work,

if the min size cannot occupy, go to create one and added into the existing work.

Reply:
fixed.